controllers/deployment.go

```diff
@@ -186,9 +186,15 @@ func buildStatefulSets(master *v1alpha1.CDAPMaster, name string, services Servic
 			continue
 		}
 
-		c := serviceContainerSpec(ss, master, dataDir, s)
+		// Merge environment variables between service spec and master spec
+		env, err := mergeEnvVars(&master.Spec, ss)
+		if err != nil {
+			return nil, err
+		}
+
+		c := serviceContainerSpec(ss, master, dataDir, s, env)
 		spec = spec.withContainer(c)
-		if err := addSystemMetricsServiceIfEnabled(spec, master, ss, dataDir, c); err != nil {
+		if err := addSystemMetricsServiceIfEnabled(spec, master, ss, dataDir, c, env); err != nil {
```

Comment:
I think this will cause the env vars for the main service to be used for the sidecar service also while ignoring the env vars of the sidecar service. If this is the cause, it's undesirable as we will not be able to reduce the java heap size for the sidecar independently of the main service.

Reply:
Fixed, seems like the original pattern of passing an env var into the serviceContainerSpec function came from an earlier prototype, and I didn't simplify. Thanks!